packages/scoped-elements/README.md

```diff
@@ -0,0 +1,232 @@
+use a polyfill, we have to scope the custom element tag names if we want to use
+different versions of them in our code. This package allows you to forget about
+how custom elements are defined, registering and scoping their tag names if it is
+necessary, avoiding the name collision problem.
+
+## Installation
+
+```bash
+npm i --save @open-wc/scoped-elements
+```
+
+## How it works
+
+`createScopedHtml` is going to define the custom element classes and creates a
+unique scoped tag for each of them if needed. In case those classes were
+previously defined then it's going to return the previous defined tag.
+
+Then, the `html` function provided by the `createScopedHtml` method is going to
+transform the template literal into another one replacing the tags used by the
+user by the ones defined by the custom elements. Finally, the transformed
+template literal is going to be processed by `lit-html`.
+
+`<my-button>${this.text}</my-button>` --> `<my-button-3>${this.text}</my-button-3>`
+
+## How to use it
+
+1. Import the `createScopedHtml` from `@open-wc/scoped-elements`.
+
+   ```js
+   import { createScopedHtml } from '@open-wc/scoped-elements';
+   ```
+
+2. Import the classes of the components you want to use.
+
+   ```js
+   import MyButton from './MyButton.js';
+   import MyPanel from './MyPanel.js';
+   ```
+
+3. Create the `html` function defining the tags you want to use for the
+   components.
+
+   ```js
+   const html = createScopedHtml({
+     'my-button': MyButton,
+     'my-panel': MyPanel,
+   });
+   ```
+
+4. Use your components in your html.
+
+   ```js
+   render() {
+     return html`
+       <my-panel class="my-panel">
+         <my-button>${this.text}</my-button>
+       </my-panel}>
+     `;
+   }
+   ```
+
+### Complete example
+
+```js
+import { css, LitElement } from 'lit-element';
+import { createScopedHtml } from '@open-wc/scoped-elements';
+import MyButton from './MyButton.js';
+import MyPanel from './MyPanel.js';
+
+const html = createScopedHtml({
+  'my-button': MyButton, // <-- bind the element with a tag in your html
+  'my-panel': MyPanel,
+});
+
+export default class MyElement extends LitElement {
+  static get styles() {
+    return css`
+      .my-panel {
+        padding: 10px;
+        background-color: grey;
+      }
+    `;
+  }
+
+  static get properties() {
+    return {
+      text: String,
+    };
+  }
+
+  render() {
+    return html`
+      <my-panel class="my-panel">
+        <my-button>${this.text}</my-button>
+      </my-panel}>
+    `;
+  }
+}
+```
+
+### What if I need to use the component tag?
+
+Sometimes we need to use the component tag name, for example, if we want to do a
+`querySelector`. In that case there is a `getTags` function available that receives
+a collection of component classes and returns a collection with the tag names
+used in their definitions.
+
+```js
+import { getTags } from '@open-wc/scoped-elements';
```

Comment:
If we want to support compiling away scoped elements, we need to think how this will interact with it

Reply:
compiling at all is not possible with this as it depends on customElements registry to look if it has a registered element for this so all this will only work in the browser if your bundler shims complet browser environment including customElement registry then sure you can :)

Reply:
to compile this away you would need something like its getUniqTagName function on the bundler side that would need to inspect the whole JS ast and do the replacements this can be done via webpack or rollup for example if you do it in the bundler this is not needed

this works only at runtime and is not bundle able or lets say compile away able.

Reply:
i have my own framework called tag-html i covered near all expectable scenarios about ssr and lit-html and any other webcomponent implamentations.

Reply:
You can compile away the runtime regexp.

Reply:
I´m resolving it because it's outdated